src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -190,15 +217,66 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
+            if (SerializedDatabaseCommands == null)
+                return;
+            if(SerializedDatabaseCommands.TryGet(nameof(DatabaseCommands), out BlittableJsonReaderArray commands) == false)
+                return;
+            
+            ClusterCommands ??= new List<ClusterTransactionDataCommand>();
+            foreach (BlittableJsonReaderObject dbCmd in commands)
+            {
+                var cmdType = dbCmd[nameof(ClusterTransactionDataCommand.Type)].ToString();
+                var docId = dbCmd[nameof(ClusterTransactionDataCommand.Id)].ToString();
+                var atomicGuardKey = GetAtomicGuardKey(docId);
+                var changeVector = dbCmd[nameof(ClusterTransactionDataCommand.ChangeVector)]?.ToString();
+                long changeVectorIndex = -1;
+                if (changeVector != null)
+                    changeVectorIndex = ChangeVectorUtils.GetEtagById(changeVector, dbTopology.ClusterTransactionIdBase64);
+                ClusterCommands.Add(new ClusterTransactionDataCommand
+                {
+                    Type = cmdType switch {
+                    
+                        nameof(CommandType.PUT) =>CommandType.CompareExchangePUT,
+                        nameof(CommandType.DELETE) => CommandType.CompareExchangeDELETE,
+                        _ => throw new ArgumentOutOfRangeException()   
+                    },
+                    Id = atomicGuardKey,
+                    Index = changeVectorIndex,
+                    Document = context.ReadObject(new DynamicJsonValue { ["Id"] = docId }, "cmp-xchg-content"),
+                    Error = $"Guard compare exchange value '{atomicGuardKey}' index does not match the transaction index's {changeVectorIndex} change vector on {docId}"
+                });
+            }
+        }
+
+        const string RvnAtomicPrefix = "rvn-atomic/";
+        public static bool IsAtomicGuardKey(string id, out string docId)
+        {
+            if (id.StartsWith(RvnAtomicPrefix) == false)
+            {
+                docId = null;
+                return false;
+            }
+
+            docId = id.Substring(RvnAtomicPrefix.Length);
+            return true;
+        }
+
+        public static string GetAtomicGuardKey(string docId)
+        {
+            return RvnAtomicPrefix + docId;
+        }
+
         public unsafe void SaveCommandsBatch(ClusterOperationContext context, long index)
         {
             if (HasDocumentsInTransaction == false)
                 return;
 
             var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
             var commandsCountPerDatabase = context.Transaction.InnerTransaction.ReadTree(ClusterStateMachine.TransactionCommandsCountPerDatabase);
-            var commands = SerializedDatabaseCommands.Clone(context);
+            var commands = context.ReadObject(SerializedDatabaseCommands, "serialized-tx-commands");
```

Comment:
Isn't effectively the same thing here? I mean Clone() vs ReadObject()

Reply:
No, `Clone` is just that, doesn't promise to do more than just copy the value.
`ReadObject` will regenerate the document, which is needed since we modify the commands.

Reply:
`Clone` will also call `ReadObject` if there are any modifications:

https://github.com/ravendb/ravendb/blob/400ef1b594b85909edfb532ef1c9c5a89e81aed5/src/Sparrow/Json/BlittableJsonReaderObject.cs#L1054-L1069

Reply:
Only on the first level, but we aren't modifying the first level.